### ISSUE TITLE
Dynamic Build Steps breaks Git Caching

### ISSUE THREAD

**Issue by sdelano** [chef/omnibus#450]
TL;DR: Dynamically defined build steps are can cause issues with git caching in omnibus. We're currently experiencing this when building Postgres, where the second build invalidates the cache and we have to start over from a much earlier point in the build than we should have to.

From a discussion with @stevendanna earlier:

```
The shasum of the software includes a digest of the builder:

  https://github.com/chef/omnibus/blob/master/lib/omnibus/software.rb#L821

The digest of the builder, includes a digest of the collected build_commands:

  https://github.com/chef/omnibus/blob/master/lib/omnibus/builder.rb#L614-L616

Part of our postgresql definition looks like this:

  # Postgres 9.2 is our "real" Postgres installation (prior versions
  # that are installed are solely to facilitate upgrades).  As a
  # result, we need to have the binaries for this version available
  # with the other binaries used by Private Chef.
  block do
    Dir.glob("#{install_dir}/embedded/postgresql/9.2/bin/*").sort.each do |bin|
      link bin, "#{install_dir}/embedded/bin/#{File.basename(bin)}"
    end
  end

Note that both `block` and `link` are part of the builder DSL.
If we calculate the shasum of the software *before* the build,
the build_commands are part of the digested:

[Builder: postgresql92] I | Incremental shasum(Execute: `./configure --prefix=/opt/postgresqltest/embedded/postgresql/9.2 --.....
[Builder: postgresql92] I | Incremental shasum(Execute: `make world -j 3') for builder: 562b3d29700088e3cdd140fb812c36a776f1ed63f8a3a8fd625a71856fc2ec50
[Builder: postgresql92] I | Incremental shasum(Execute: `make install-world -j 3') for builder: ff6b1ed515a24c5497365f82d21a41eb6a9add0aa2157bb48a25745e38707a8b
[Builder: postgresql92] I | Incremental shasum(<Dynamic Ruby block>) for builder: 842bdd5519e0f2020e5a69e8d0f4b218bbc88e1ec8e80306047d07258a96ba15

However, if we calculate the shasum *after* the build, the
dynamic resource block has been executed and all of those link
statements have been added to the build_commands:

[Builder: postgresql92] I | Incremental shasum(Execute: `./configure --prefix=/opt/postgresqltest/embedded/postgresql/9.2 --with-.....
[Builder: postgresql92] I | Incremental shasum(Execute: `make world -j 3') for builder: 562b3d29700088e3cdd140fb812c36a776f1ed63f8a3a8fd625a71856fc2ec50
[Builder: postgresql92] I | Incremental shasum(Execute: `make install-world -j 3') for builder: ff6b1ed515a24c5497365f82d21a41eb6a9add0aa2157bb48a25745e38707a8b
[Builder: postgresql92] I | Incremental shasum(<Dynamic Ruby block>) for builder: 842bdd5519e0f2020e5a69e8d0f4b218bbc88e1ec8e80306047d07258a96ba15
[Builder: postgresql92] I | Incremental shasum(link `/opt/postgresqltest/embedded/postgresql/9.2/bin/clusterdb' to `/opt/postgresqltes...
[Builder: postgresql92] I | Incremental shasum(link `/opt/postgresqltest/embedded/postgresql/9.2/bin/createlang' to `/opt/...
[Builder: postgresql92] I | Incremental shasum(link `/opt/postgresqltest/embedded/postgresql/9.2/bin/....
[Builder: postgresql92] I | Incremental shasum(link `/opt/postgresqltest/embedded/postgresql/9.2/bin/dropdb' to `/opt/postgresqltest/embe..

Thus, the following can happen:

1) First run, something *before* postgresql dirties the
cache. This causes us to start building before we ever call
software#shasum:

https://github.com/chef/omnibus/blob/master/lib/omnibus/software.rb#L763-L767

2) At the end of the build, we update the gitcache, calling
shasum for the first time and building the digest with all the link resources as
shown above.

3) On the second run, we should be able to restore from the
cache, but we call software#shasum before the build, which means
the link resources haven't been added to the build_commands and
thus we calculate a different shasum.
```

There are a couple of ways that we can approach a fix for this:

1. We could convert the `block` in question to a bash command that does the equivalent. Said command would result in the icallsame shasum for the build instructions every time. This doesn't get around the issue that we can dynamically alter the build instructions based on the state of the system.
2. Steven has suggested that we move the `shasum` portion of the caching to take place before the build is execute, so that any side effects of the build are not present.
3. My concern is that we're even allowing DSL methods from within the `block` resource. Is that something that we want to do? How common is it?

**Comment by schisamo** [comment]
After an initial read of the issue I think __# 2__ is the right hard thing to do as it will ensure software definitions other than `postgresql92` will receive a fix. 

I'll have to think more about __# 3__. There are situations where taking advantage of the pure-Ruby DSL methods in build blocks is the right thing to do for things like readability or platform portability (maybe not in this example though).

**Comment by stevendanna** [comment]
I've filed a PR based on solution (2).

> My concern is that we're even allowing DSL methods from within the block resource. Is that something that we want to do? How common is it?

Regarding (3), the only software definition in omnibus-software that uses DSL methods inside blocks is the "preparation" software. Only the postgresql92 definition in the opscode-omnibus repo.  In the omnibus-chef repository a couple of definitions use DSL methods, one expects to be able to access the software definition directly from inside that block.

That said, after thinking about it this morning, restricting DSL methods inside the block resource might only simplify the situation for us (as omnibus developers) but not actually do much for correctness.  The shasum of a builder would still be mostly a lie since all unnamed blocks look the same to the builder's shasum method.  This problem is mitigated for the user already since we include the actual text of the software definition in the software digest (which is what is ultimately used for caching):

https://github.com/chef/omnibus/blob/master/lib/omnibus/software.rb#L827

Further, since we allow arbitrary ruby, it will always be possible to construct a software definition that can produce a different build based on system state.

As a note: I noticed while writing this that we don't include the version of omnibus anywhere in the data that contributes to the git tag.  Does this seem incorrect to anyone else?
